### ISSUE TITLE
Invalid configuration `arm64-apple-darwin20.3.0': machine `arm64-apple' not recognized

### ISSUE THREAD

**Issue by milosivanovic** [pyca/pynacl#657]
I am not able to build pynacl-1.4.0/main from source (for the arm64 version of python3.8) on an Apple Silicon / M1 machine.
My terminal and all utilities are natively installed from the M1 version of brew. Rosetta 2 is not being used.

Here is the full build output:

```
(env) milos@air:~/d/pynacl$ python3 setup.py install
running install
running bdist_egg
running egg_info
creating src/PyNaCl.egg-info
writing src/PyNaCl.egg-info/PKG-INFO
writing dependency_links to src/PyNaCl.egg-info/dependency_links.txt
writing requirements to src/PyNaCl.egg-info/requires.txt
writing top-level names to src/PyNaCl.egg-info/top_level.txt
writing manifest file 'src/PyNaCl.egg-info/SOURCES.txt'
reading manifest file 'src/PyNaCl.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no previously-included files matching '__pycache__/*' found anywhere in distribution
warning: no previously-included files found matching '.travis.yml'
warning: no previously-included files found matching '.travis'
warning: no previously-included files found matching '.travis/install.sh'
warning: no previously-included files found matching '.travis/run.sh'
warning: no previously-included files found matching '.readthedocs.yml'
warning: no previously-included files found matching '.coveragerc'
warning: no previously-included files found matching 'codecov.yml'
warning: no previously-included files found matching 'release.py'
warning: no previously-included files found matching 'src/libsodium*.zip'
writing manifest file 'src/PyNaCl.egg-info/SOURCES.txt'
installing library code to build/bdist.macosx-11-arm64/egg
running build_clib
checking build system type... Invalid configuration `arm64-apple-darwin20.3.0': machine `arm64-apple' not recognized
configure: error: /bin/sh /Users/milos/d/pynacl/src/libsodium/build-aux/config.sub arm64-apple-darwin20.3.0 failed
Traceback (most recent call last):
  File "setup.py", line 216, in <module>
    setup(
  File "/Users/milos/d/testproj/env/lib/python3.8/site-packages/setuptools/__init__.py", line 153, in setup
    return distutils.core.setup(**attrs)
  File "/opt/homebrew/Cellar/python@3.8/3.8.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/opt/homebrew/Cellar/python@3.8/3.8.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/opt/homebrew/Cellar/python@3.8/3.8.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/Users/milos/d/testproj/env/lib/python3.8/site-packages/setuptools/command/install.py", line 67, in run
    self.do_egg_install()
  File "/Users/milos/d/testproj/env/lib/python3.8/site-packages/setuptools/command/install.py", line 109, in do_egg_install
    self.run_command('bdist_egg')
  File "/opt/homebrew/Cellar/python@3.8/3.8.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/opt/homebrew/Cellar/python@3.8/3.8.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/Users/milos/d/testproj/env/lib/python3.8/site-packages/setuptools/command/bdist_egg.py", line 163, in run
    self.run_command('build_clib')
  File "/opt/homebrew/Cellar/python@3.8/3.8.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/opt/homebrew/Cellar/python@3.8/3.8.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "setup.py", line 176, in run
    subprocess.check_call(
  File "/opt/homebrew/Cellar/python@3.8/3.8.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/Users/milos/d/pynacl/src/libsodium/configure', '--disable-shared', '--enable-static', '--disable-debug', '--disable-dependency-tracking', '--with-pic', '--prefix', '/Users/milos/d/pynacl/build/temp.macosx-11-arm64-3.8']' returned non-zero exit status 1.
```

I tried 1.4.0 and master, both result in the same configure error.

As far as I'm aware, `config.sub` generated by autoconf. Would this be a pynacl issue or autoconf issue, or something else? From https://github.com/pyca/pynacl/issues/654#issuecomment-763086607 it sounds like you have an M1 machine yourself, do you remember how you were able to get this compiled on it?

**Comment by milosivanovic** [comment]
I was able to get it to build successfully by replacing the contents of the `src/libsodium/build-aux/config.sub` file with just `arm64-apple-darwin`.

It seems the solution here is to modify that file properly to support this new machine type.

**Comment by reaperhulk** [comment]
I haven't tried this on Python 3.8, but using homebrew Python 3.9 or python.org Python 3.9 I was able to compile with no issues. When I look at the build system type that's detected I see `arm-apple-darwin20.3.0` though, not `arm64-apple-darwin20.3.0`.

**Comment by reaperhulk** [comment]
config.sub is provided by libsodium; if you have a chance please try compiling the latest libsodium-1.0.18-stable from libsodium.org. If that succeeds then libsodium has already handled whatever is making your autotools/configure/make behave differently (perhaps different versions?).

**Comment by milosivanovic** [comment]
Yes, libsodium 1.0.18-stable works, but the config.sub file has a different timestamp than the file in the pynacl repo.

When I replace `src/libsodium/build-aux/config.sub` with the one from libsodium-1.0.18-stable, then pynacl also builds successfully. This means the version of config.sub does not match the one from libsodium, and I can already tell from the differing timestamp at the top of the file that it does not match (timestamp='2019-05-23' for pynacl/main, vs timestamp='2021-01-08' for libsodium-1.0.18-stable)

**Comment by reaperhulk** [comment]
Yeah, I didn't expect them to be the same. So it looks like we probably need to upgrade pynacl to the latest libsodium stable. (libsodium has an unusual versioning scheme where 1.0.18 was released long ago and then there are daily -stable "releases" that update it but no 1.0.19 has occurred).

**Comment by milosivanovic** [comment]
That should be doable then! This workaround isn't terrible in the meantime for anyone else that might stumble across this issue.

What I don't understand is why this would work for you on Python 3.9 (which I haven't tried)?

**Comment by reaperhulk** [comment]
It shouldn't be related to your Python. The way that information is determined is via `build-aux/config.guess`. I am unsure what changed in the detection there that would result in any differences between two M1 macs though.

**Comment by milosivanovic** [comment]
I ran into this same issue building another python package (gevent), but now I'm confused because at this rate it seems like a whole lot of packages will be broken. The target clang provides when queried on M1 is likely to be missing support in many projects that use an older `config.sub`:

```
$ clang -v 2>&1 | grep Target
Target: arm64-apple-darwin20.3.0
```

In terms of libsodium, the additional lines in config.sub (pulled from https://github.com/jedisct1/libsodium/blob/master/autogen.sh#L112) that allowed pynacl to build successfully on my machine is:

```
	arm64-*)
		cpu=aarch64
		;;

```

which, when missing, causes `Invalid configuration` errors like the one in my first comment.

For example, with the `gevent` package I was trying to build, it fails with:

```
running build_ext
  generating cffi module 'build/temp.macosx-11-arm64-3.8/gevent.libuv._corecffi.c'
  creating build/temp.macosx-11-arm64-3.8
  Running '(cd  "/private/var/folders/y5/vfyf5lqj655g_zbm8pvdgmkr0000gn/T/pip-install-7vfntgw4/gevent_1d7b0d201cde413b98ff1d26002d091a/deps/libev"  && sh ./configure -C > configure-output.txt )' in /private/var/folders/y5/vfyf5lqj655g_zbm
8pvdgmkr0000gn/T/pip-install-7vfntgw4/gevent_1d7b0d201cde413b98ff1d26002d091a
  Invalid configuration `arm64-apple-darwin20.3.0': machine `arm64-apple' not recognized
  configure: error: /bin/sh ./config.sub arm64-apple-darwin20.3.0 failed
```

According to an unrelated project (https://camcops.readthedocs.io/en/latest/autodoc/tablet_qt/tools/build_qt.py.html) I see this:

```
    @property
    def apple_cpu_name_for_triplet(self) -> str:
        """
        CPU name to make a cpu-vendor-os triplet.

        See :meth:`apple_arch_name`.

        Note that "arm64" is a valid architecture but fails here (e.g.
        SQLCipher ``configure``) with ``Invalid configuration
        'arm64-apple-darwin': machine 'arm64-apple' not recognized". The
        solution is apparently to use ``arm-apple-darwin`` but pass ``-m64`` to
        clang, and/or pass ``-arch arm64`` to clang (the latter is more
        plausible to me); see
        https://github.com/tpoechtrager/cctools-port/issues/6.
        """  # noqa
        if self.cpu_x86_64bit_family:
            return "x86_64"
        elif self.cpu_x86_32bit_family:
            return "i386"
        elif self.cpu_arm_family:
            # See above
            return "arm"
        else:
            raise ValueError(f"apple_cpu_name_for_triplet(): "
                             f"Unknown CPU {self.cpu}")
```

which makes me wonder what the right solution for this overall issue is really supposed to be. Should we be supporting arm64 in config.sub? Or should we be modifying the target to be `arm` and then use -m64 when compiling?

I haven't seen any reports about people compiling packages with pip and running into these compilation issues, so either I'm the first to notice, or I am doing something wrong...

**Comment by lloeki** [comment]
> I haven't seen any reports about people compiling packages with pip and running into these compilation issues, so either I'm the first to notice, or I am doing something wrong...

I'm hitting this over here. (with an ARM nixpkgs):

```
$ cat shell.nix 
{ pkgs ? import <nixpkgs> {} }:

pkgs.mkShell {
  nativeBuildInputs = [ pkgs.pkg-config ];
  buildInputs = [
    pkgs.python3
    pkgs.libjpeg
    pkgs.autoconf
    pkgs.libtiff
    pkgs.openssl
    pkgs.libffi

    pkgs.nginx

    pkgs.zlib
    pkgs.libwebp
    pkgs.lcms2
    pkgs.libimagequant
  ]
  ++ pkgs.lib.optionals pkgs.stdenv.isDarwin (with pkgs.darwin.apple_sdk.frameworks; [
    Cocoa CoreServices CoreData Foundation
  ]);
}
$ nix-shell
nix $ python3 -m venv venv
nix $ ./venv/bin/python3 -m pip install pynacl==1.4.0
...
  checking build system type... Invalid configuration `arm64-apple-darwin20.5.0': machine `arm64-apple' not recognized
  configure: error: /nix/store/zj3mf2cki8hcs5qmkdns85nkncdqsmjb-bash-4.4-p23/bin/bash /private/tmp/pip-install-srdg90k8/pynacl/src/libsodium/build-aux/config.sub arm64-apple-darwin20.5.0 failed
...
```

**Comment by 0x91** [comment]
> > I haven't seen any reports about people compiling packages with pip and running into these compilation issues, so either I'm the first to notice, or I am doing something wrong...
> 
> I'm hitting this over here. (with an ARM nixpkgs):
> 
> ```
> $ cat shell.nix 
> { pkgs ? import <nixpkgs> {} }:
> 
> pkgs.mkShell {
>   nativeBuildInputs = [ pkgs.pkg-config ];
>   buildInputs = [
>     pkgs.python3
>     pkgs.libjpeg
>     pkgs.autoconf
>     pkgs.libtiff
>     pkgs.openssl
>     pkgs.libffi
> 
>     pkgs.nginx
> 
>     pkgs.zlib
>     pkgs.libwebp
>     pkgs.lcms2
>     pkgs.libimagequant
>   ]
>   ++ pkgs.lib.optionals pkgs.stdenv.isDarwin (with pkgs.darwin.apple_sdk.frameworks; [
>     Cocoa CoreServices CoreData Foundation
>   ]);
> }
> $ nix-shell
> nix $ python3 -m venv venv
> nix $ ./venv/bin/python3 -m pip install pynacl==1.4.0
> ...
>   checking build system type... Invalid configuration `arm64-apple-darwin20.5.0': machine `arm64-apple' not recognized
>   configure: error: /nix/store/zj3mf2cki8hcs5qmkdns85nkncdqsmjb-bash-4.4-p23/bin/bash /private/tmp/pip-install-srdg90k8/pynacl/src/libsodium/build-aux/config.sub arm64-apple-darwin20.5.0 failed
> ...
> ```

likewise on python 3.8

**Comment by maxmouchet** [comment]
@lloeki @0x91 you're probably using coreutil's uname, which returns `arm64`. macOS native uname (`/usr/bin/uname`) returns `arm`, which works correctly. Make sure that macOS uname is first on the path:
```
/usr/bin/uname -p # arm
/nix/store/czksxcbg7w21f3xb7v18vj58izwfhhwq-coreutils-8.32/bin/uname -p # arm64
```
If needed, you can use `coreutils-prefixed` which prefixes every binary with `g` (e.g. `guname`).

**Comment by lloeki** [comment]
Looks like so, but it's being exposed by Nix without me specifying it (probably pulled in as a dependency and exposed to the shell for consistency):

```
$ which uname
/nix/store/vby6bv9y59a6p44cjp7cw3pbw6y04x68-coreutils-8.32/bin/uname
```

Hacking `PATH` to have `/usr/bin/uname` first sounds very brittle, especially given Nix's design where `/nix/store/vby6bv9y59a6p44cjp7cw3pbw6y04x68-coreutils-8.32/bin/uname` could be referenced with an absolute path.

**Comment by jace** [comment]
This worked for me:

```
PATH="/usr/bin:$PATH" pip install pynacl
```

**Comment by reaperhulk** [comment]
I would expect this is generally resolved by the 1.5.0 release. Please open a new issue if you see it in 1.5.0+.
